ephios/plugins/basesignup/signup/section_based.py

```diff
@@ -0,0 +1,319 @@
+            formset.save()
+            return redirect(self.object.event.get_absolute_url())
+        return self.get(request, *args, **kwargs, formset=formset)
+
+    def get_context_data(self, **kwargs):
+        kwargs.setdefault("formset", self.get_formset())
+        kwargs.setdefault("states", AbstractParticipation.States)
+        kwargs.setdefault("sections", self.object.signup_method.configuration.sections)
+        return super().get_context_data(**kwargs)
+
+
+class SectionForm(forms.Form):
+    title = forms.CharField(label=_("Title"), required=True)
+    qualifications = forms.ModelMultipleChoiceField(
+        label=_("Required Qualifications"),
+        queryset=Qualification.objects.all(),
+        widget=Select2MultipleWidget,
+        required=False,
+    )
+    min_count = forms.IntegerField(label=_("min amount"), min_value=1, required=True)
+    uuid = forms.CharField(widget=forms.HiddenInput, required=False)
+
+    def clean_uuid(self):
+        return self.cleaned_data.get("uuid") or uuid.uuid4()
+
+
+SectionsFormset = forms.formset_factory(
+    SectionForm, can_delete=True, min_num=1, validate_min=1, extra=0
+)
+
+
+class SectionBasedConfigurationForm(forms.Form):
+    def __init__(self, data=None, **kwargs):
+        super().__init__(data, **kwargs)
+        self.sections_formset = SectionsFormset(
+            data=data,
+            initial=self.initial.get("sections", list()),
+            prefix="sections",
+        )
+
+    def clean_sections(self):
+        if not self.sections_formset.is_valid():
+            raise ValidationError(_("The sections aren't configured correctly."))
+
+        sections = [
+            {
+                key: form.cleaned_data[key]
+                for key in ("title", "qualifications", "min_count", "uuid")
+            }
+            for form in self.sections_formset
+        ]
+        return sections
+
+
+class SectionSignupForm(forms.Form):
+    section = forms.ChoiceField(
+        label=_("Preferred Section"),
+        widget=forms.RadioSelect,
+        # choices are set as (uuid, title) of section
+    )
+
+
+class SectionBasedSignupView(FormView, BaseSignupView):
+    template_name = "basesignup/section_based/signup.html"
+
+    def get_form(self, form_class=None):
+        form = SectionSignupForm(self.request.POST)
+        form.fields["section"].choices = [
+            (section["uuid"], section["title"])
+            for section in sections_participant_qualifies_for(
+                self.method.configuration.sections, self.request.user.as_participant()
+            )
+        ]
+        return form
+
+    def get_context_data(self, **kwargs):
+        kwargs.setdefault("shift", self.shift)
+        kwargs.setdefault(
+            "unqualified_sections",
+            [
+                section["title"]
+                for section in self.method.configuration.sections
+                if section
+                not in sections_participant_qualifies_for(
```

Comment:
is `sections_participant_qualifies_for` executed for every iteration of the loop?

Reply:
yes. Will fix that :)